src/app/compose/recipients.service.ts

```diff
@@ -48,48 +50,70 @@ export class RecipientsService {
 
-                    contact.emails.forEach(email => {
-                        const recipientString = `"${contact.first_and_last_name()}" <${email.value}>`;
-                        recipientsMap[email.value] = Recipient.fromContact(contact, email.value);
-                    });
+            const categories = {};
+            const groups     = [];
+            contacts.forEach(contact => {
+                if (contact.kind === ContactKind.GROUP) {
+                    groups.push(contact);
+                    return;
+                }
 
-                    contact.categories.forEach(category => {
-                        if (!categories[category]) {
-                            categories[category] = [];
-                        }
-                        categories[category].push(contact);
+                contact.emails.forEach(email => {
+                    this.recipientsUpdating.push({
+                        'origin': 'contacts',
+                        'uniqueKey': email.value,
+                        'recipient': Recipient.fromContact(contact, email.value)
                     });
                 });
 
-                const result = Object.keys(recipientsMap).map(mailaddr => recipientsMap[mailaddr]);
+                contact.categories.forEach(category => {
+                    if (!categories[category]) {
+                        categories[category] = [];
+                    }
+                    categories[category].push(contact);
+                });
+            });
 
-                for (const category of Object.keys(categories)) {
-                    result.unshift(Recipient.fromCategory(category, categories[category]));
-                }
+            for (const category of Object.keys(categories)) {
+                this.recipientsUpdating.push({
+                    'origin': 'contacts',
+                    'uniqueKey': category,
+                    'recipient': Recipient.fromCategory(category, categories[category])});
+            }
 
-                Promise.all(
-                    groups.map(g => this.recipientFromGroup(g))
-                ).then(
-                    recipients => this.recipients.next(result.concat(recipients))
-                ).catch(
-                    () => this.recipients.next(result)
-                );
-            });
+            Promise.all(
+                groups.map(g => this.recipientFromGroup(g))
+            ).then((updateGroups) => {
+                this.updateRecipients(updateGroups);
+            }).catch(
+                () => this.recipients.next([])
+            );
+        });
+    }
+
+    private updateRecipients(groups: Recipient[] = []) {
+        const uniqueRecipients: {[uniqueKey: string]: Recipient} = {};
+        this.recipientsUpdating.forEach(entry => {
+            if (!uniqueRecipients[entry['uniqueKey']] ||
+                entry['origin'] === 'searchindex') {
```

Comment:
These are set to either `'contacts'` or `'searchindex'` in other parts of the code – so this part of the `if` will always eval to false, I think.

Perhaps it'd be better to switch to an enum for these, to avoid typos like this :)

It also seems like in its current form (if corrected), it'd make searchindex win over contacts when updating the list – is that intentional, and/or covered by our tests?

Reply:
Oops, yup that was me being a nitwit (tests pass tho!) .. enum sounds like a good idea!